test/test_foreach.py

```diff
@@ -0,0 +1,102 @@
+import torch
+import torch.cuda
+from torch.testing._internal.common_utils import TestCase, run_tests
+from torch.testing._internal.common_device_type import instantiate_device_type_tests, dtypes
+
+class TestForeach(TestCase):
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_add_scalar_with_same_size_tensors(self, device, dtype):
+        N = 20
+        H = 20
+        W = 20
+        tensors = []
+        for _ in range(N):
+            tensors.append(torch.zeros(H, W, device=device, dtype=dtype))
+
+        res = torch._foreach_add(tensors, 1)
+        for t in res:
+            if dtype == torch.bool:
+                dtype = torch.int64
+            self.assertEqual(t, torch.ones(H, W, device=device, dtype=dtype))
+
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_add_scalar_with_different_size_tensors(self, device, dtype):
+        N = 20
+        H = 20
+        W = 20
+
+        tensors = []
+        size_change = 0
+        for _ in range(N):
+            tensors.append(torch.zeros(H + size_change, W + size_change, device=device, dtype=dtype))
+            size_change += 1
+
+        res = torch._foreach_add(tensors, 1)
+
+        size_change = 0
+        for t in res: 
+            if dtype == torch.bool:
+                dtype = torch.int64
+            self.assertEqual(t, torch.ones(H + size_change, W + size_change, device=device, dtype=dtype))
+            size_change += 1
+
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_add_scalar_with_empty_list(self, device, dtype):
+        tensors = []
+        with self.assertRaises(RuntimeError):
+            torch._foreach_add(tensors, 1)
```

Comment:
how do you expect this to work with an optimizer with no parameters?  I would expect this to generalize from adding a scalar to a tensor with no elements, i.e. do nothing.

```
>>> torch.randn([0]) + 5
tensor([])
```